body/stretch_body/hello_utils.py

```diff
@@ -113,32 +114,80 @@ def pretty_print_dict(title, d):
+class LoopStats():
+    """
+    Track timing statistics for control loops
+    """
+    def __init__(self,loop_name,target_loop_rate):
+        self.loop_name=loop_name
+        self.target_loop_rate=target_loop_rate
+        self.ts_loop_start=None
+        self.ts_loop_end=None
+        self.status={'loop_rate_hz':0, 'loop_rate_avg_hz':0, 'loop_rate_min_hz':10000000, 'loop_rate_max_hz':0,'loop_rate_std':0, 'execution_time_ms':0, 'loop_warns':0}
+        self.logger = logging.getLogger()
+        self.n_log=100
+        self.log_idx=0
+        self.rate_log=None
+        self.log_rate_hz=1.0
+        self.warned_yet=False
+        self.sleep_time_s =0.001
+        self.loop_cycles=0
 
     def pretty_print(self):
-        print('Timer Stat -- Avg: ', str(self.av), 'Max: ', str(self.mx))
+        print('--------- TimingStats %s -----------'%self.loop_name)
+        print('Target rate: %f'%self.target_loop_rate)
+        print('Current rate (Hz): %f' % self.status['loop_rate_hz'])
+        print('Average rate (Hz): %f' % self.status['loop_rate_avg_hz'])
+        print('Min rate (Hz): %f' % self.status['loop_rate_min_hz'])
+        print('Max rate (Hz): %f' % self.status['loop_rate_max_hz'])
+        print('Current execution time (ms): %f' % self.status['execution_time_ms'])
+        print('Execution time supports rate of (Hz) %f'%(1000.0/self.status['execution_time_ms']))
+        print('Warnings: %d out of %d'%(self.status['loop_warns'],self.loop_cycles))
+
+    def mark_loop_start(self):
+        self.ts_loop_start=time.time()
+
+    def mark_loop_end(self):
+        self.loop_cycles+=1
+        #First two cycles initialize vars / log
+        if not self.ts_loop_start:
+            return
+        if self.ts_loop_end is None:
+            self.ts_loop_end=time.time()
+            return
+        end_last=self.ts_loop_end
+        self.ts_loop_end=time.time()
+        self.status['execution_time_ms']=(self.ts_loop_end-self.ts_loop_start)*1000
+        self.status['loop_rate_hz']=1.0/(self.ts_loop_end-end_last)
+        self.status['loop_rate_min_hz']=min(self.status['loop_rate_hz'], self.status['loop_rate_min_hz'])
+        self.status['loop_rate_max_hz'] = max(self.status['loop_rate_hz'], self.status['loop_rate_max_hz'])
+        if type(self.rate_log)==type(None):
+            self.rate_log=numpy.array([self.status['loop_rate_hz']] * self.n_log)
+        self.rate_log[self.log_idx]=self.status['loop_rate_hz']
+        self.log_idx=(self.log_idx+1)%self.n_log
+        self.status['loop_rate_avg_hz']=numpy.average(self.rate_log)
+
+        self.sleep_time_s = (1 / self.target_loop_rate) - (self.status['execution_time_ms'] / 1000)
+        if self.sleep_time_s < .001:
+            self.status['loop_warns'] += 1
+            if not self.warned_yet:
+                self.warned_yet=True
+                self.logger.warning('Target loop rate for %s not possible. Capable of %.2f ms' % (self.loop_name,(1000.0/self.status['execution_time_ms'])))
```

Comment:
The warning should include target and capable loop rate in Hz. It's difficult to interpret milliseconds.

Reply:
Also, let's make this a debug statement, so it doesn't confuse users. At the moment, I'm seeing it when homing the robot.